pkg/disttask/framework/scheduler/balancer.go

```diff
@@ -0,0 +1,192 @@
+			return
+		}
+	}
+}
+
+func (b *balancer) balanceSubtasks(ctx context.Context, sch Scheduler, managedNodes []string) error {
+	task := sch.GetTask()
+	eligibleNodes, err := getEligibleNodes(ctx, sch, managedNodes)
+	if err != nil {
+		return err
+	}
+	if len(eligibleNodes) == 0 {
+		return errors.New("no eligible nodes to balance subtasks")
+	}
+	return b.doBalanceSubtasks(ctx, task.ID, eligibleNodes)
+}
+
+func (b *balancer) doBalanceSubtasks(ctx context.Context, taskID int64, eligibleNodes []string) (err error) {
+	subtasks, err := b.taskMgr.GetActiveSubtasks(ctx, taskID)
+	if err != nil {
+		return err
+	}
+	if len(subtasks) == 0 {
+		return nil
+	}
+
+	// balance subtasks only to nodes with enough slots, from the view of all
+	// managed nodes, subtasks of task might not be balanced.
+	adjustedNodes := filterNodesWithEnoughSlots(b.currUsedSlots, b.slotMgr.getCapacity(),
+		eligibleNodes, subtasks[0].Concurrency)
+	if len(adjustedNodes) == 0 {
+		// no node has enough slots to run the subtasks, skip balance and skip
+		// update used slots.
+		return nil
+	}
+	adjustedNodeMap := make(map[string]struct{}, len(adjustedNodes))
+	for _, n := range adjustedNodes {
+		adjustedNodeMap[n] = struct{}{}
+	}
+
+	defer func() {
+		if err == nil {
+			b.updateUsedNodes(subtasks)
+		}
+	}()
+
+	averageSubtaskCnt := (len(subtasks) + len(adjustedNodes) - 1) / len(adjustedNodes)
+	executorSubtasks := make(map[string][]*proto.Subtask, len(adjustedNodes))
+	executorPendingCnts := make(map[string]int, len(adjustedNodes))
+	for _, node := range adjustedNodes {
+		executorSubtasks[node] = make([]*proto.Subtask, 0, averageSubtaskCnt)
+	}
+	for _, subtask := range subtasks {
+		// put running subtask in the front of slice.
+		// if subtask fail-over, it's possible that there are multiple running
+		// subtasks for one task executor.
+		if subtask.State == proto.TaskStateRunning {
+			executorSubtasks[subtask.ExecID] = append([]*proto.Subtask{subtask}, executorSubtasks[subtask.ExecID]...)
+		} else {
+			executorSubtasks[subtask.ExecID] = append(executorSubtasks[subtask.ExecID], subtask)
+			executorPendingCnts[subtask.ExecID]++
+		}
+	}
+
+	subtasksNeedSchedule := make([]*proto.Subtask, 0)
+	for k, v := range executorSubtasks {
+		if _, ok := adjustedNodeMap[k]; !ok {
+			// dead node or not have enough slots
+			subtasksNeedSchedule = append(subtasksNeedSchedule, v...)
+			delete(executorSubtasks, k)
+			continue
+		}
+		if len(v) > averageSubtaskCnt {
+			// running subtasks are never balanced.
+			cnt := min(executorPendingCnts[k], len(v)-averageSubtaskCnt)
+			subtasksNeedSchedule = append(subtasksNeedSchedule, v[len(v)-cnt:]...)
+			executorSubtasks[k] = v[:len(v)-cnt]
+		}
+	}
+	if len(subtasksNeedSchedule) == 0 {
+		return nil
+	}
+
+	fillIdx := 0
+	for _, node := range adjustedNodes {
+		sts := executorSubtasks[node]
+		for i := len(sts); i < averageSubtaskCnt && fillIdx < len(subtasksNeedSchedule); i++ {
```

Comment:
from @ywqzzy : suppose there are 9 subtasks to balance on 4 nodes, we should have 3/2/2/2, not 3/3/3

Reply:
subtasks: []*proto.Subtask{
				{ID: 1, ExecID: "tidb1", Concurrency: 16, State: proto.TaskStatePending},
				{ID: 2, ExecID: "tidb1", Concurrency: 16, State: proto.TaskStatePending},
				{ID: 3, ExecID: "tidb1", Concurrency: 16, State: proto.TaskStatePending},
				{ID: 4, ExecID: "tidb1", Concurrency: 16, State: proto.TaskStatePending},
				{ID: 5, ExecID: "tidb2", Concurrency: 16, State: proto.TaskStatePending},
				{ID: 6, ExecID: "tidb2", Concurrency: 16, State: proto.TaskStatePending},
				{ID: 7, ExecID: "tidb2", Concurrency: 16, State: proto.TaskStatePending},
				{ID: 8, ExecID: "tidb2", Concurrency: 16, State: proto.TaskStatePending},
				{ID: 9, ExecID: "tidb2", Concurrency: 16, State: proto.TaskStatePending},
			},
			eligibleNodes: []string{"tidb1", "tidb2", "tidb3", "tidb4"},
			initUsedSlots: map[string]int{"tidb2": 0, "tidb3": 0, "tidb1": 0, "tidb4": 0},
}
case here.